miniapps/meshing/ext-mesh-mapping.cpp

```diff
@@ -0,0 +1,310 @@
+   //The partition is the MPI Rank that each element lives on.
+   Array<int> partition({3,3,1,2,0});
+   ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh, partition.GetData());
+
+   //Compute the mappings between the local element ids on each processor in pmesh
+   //and the original global IDs in the mesh.  This can be computed from the element
+   //partition array.
+   Array<int> pmesh_to_mesh_emap;
+   create_pmesh_to_mesh_emaps(partition, pmesh, pmesh_to_mesh_emap);
+
+   //Now compute the mappings between the local vertex ids on each processor in pmesh
+   //and the global ids in mesh.
+   Array<int> pmesh_to_mesh_vmap;
+   create_pmesh_to_mesh_vmaps(pmesh, mesh, pmesh_to_mesh_emap, pmesh_to_mesh_vmap);
+
+   //Now compose the maps to create a final map between the local vertex numbering in the
+   //pmesh object on this processor and the original dmesh vertex numbering
+   Array<int> final_vmap(pmesh->GetNV());
+   for (int local_vi = 0; local_vi < pmesh->GetNV(); ++local_vi)
+   {
+      final_vmap[local_vi] = mesh_to_dmesh_vmap[pmesh_to_mesh_vmap[local_vi]];
+   }
+   print_pmesh_verts(pmesh,
+                     final_vmap);           //Print the vertices with global IDs from the final_vmap
+}
+
+// Build up an MFEM Mesh from the data in the Dummy Mesh.  Since we are
+// building the vertex and element lists in the same order as we found them
+// in the dmesh, the element id and vertex id mappings will be the identity maps.
+Mesh *build_mfem_mesh(DummyMesh *dmesh)
+{
+   //Initilize the the dimension and memory for the mesh
+   Mesh *mesh = new Mesh(2,    // The dimension of the mesh
+                         dmesh->num_vertices,
+                         dmesh->num_elements,
+                         dmesh->num_belements,
+                         2    // The dimension of the space the mesh lives in (different for surface meshes)
+                        );
+
+
+   //Add the vertices into the mesh in the same order as they are in the dmesh
+   for (int vid = 0; vid < dmesh->num_vertices; ++vid)
+   {
+      mesh->AddVertex(dmesh->V[vid].x, dmesh->V[vid].y);
+   }
+
+   //Add the elements into the mesh in the same order as they are in the dmesh
+   //The dmesh is purely quads, but if it had other types we would use
+   //the other mesh.Add* methods here as well.
+   for (int eid = 0; eid < dmesh->num_elements; ++eid)
+   {
+      std::array<int,4> vid = dmesh->E[eid].vertex_ids;
+
+      //The order of vertices in the MFEM quad elements is different from the
+      //order of the vertices in out DummyMesh format so we must permute them
+      //here.  See the ref-*.mesh files in mfem/data to establish to ordering
+      //of the vertices in the MFEM elements.
+      //  Dummy    MFEM
+      //  2--3     3--2
+      //  |  |     |  |
+      //  0--1     0--1
+      mesh->AddQuad(vid[0], vid[1], vid[3], vid[2]);
+   }
+
+   //Add the boundary elements into the mesh in the same order they are in the dmesh
```

Comment:
Technically your `DummyMesh` stores some boundary segments with a different orientation than an `mfem::Mesh` would. If the `mfem::Mesh` were conforming we would see warnings from an orientation check during `Mesh::FinalizeMesh`. Since we're using an `mfem::NCMesh` here the orientations are quietly altered. Consequently, if the `mfem::Mesh` object is written to disk its boundary elements, or some of them anyway, will have different orientation from the original mesh. This might lead to incorrect assumptions for some users who might expand on this example code.

I would suggest that you expand this comment to describe the orientation of boundary elements used in MFEM. I think we should at least give users a hint here that they should put some thought into how MFEM's assumptions may differ from those of their native mesh format.

Reply:
Oh I didn't catch that.  That is exactly the sort of thing this miniapp is supposed to help users sort out.  I'll look into it and make sure it is clear in this example.